crates/fuzzing/src/oracles/dummy/wencoder_generator.rs

```diff
@@ -0,0 +1,198 @@
+    pub fn finish(self) -> Vec<u8> {
+        self.module.finish()
+    }
+    pub fn import(&mut self, ty: &ImportType<'_>) {
+        let mut imports = ImportSection::new();
+        imports.import(ty.module(), ty.name(), extern_to_entity(&ty.ty()));
+        self.module.section(&imports);
+    }
+    pub fn export(&mut self, ty: &ExportType<'_>) {
+        let nth = self.next(&ty.ty());
+        let section_name = format!("item{}", nth);
+        let mut exports = ExportSection::new();
+        let item_ty = ty.ty();
+        self.item(&item_ty);
+        let export = extern_to_export(&item_ty, |_| nth);
+        exports.export(&section_name, export);
+        self.module.section(&exports);
+    }
+    fn item(&mut self, ty: &ExternType) {
+        match ty {
+            ExternType::Memory(mem) => {
+                let mut memories = MemorySection::new();
+                memories.memory(wasm_encoder::MemoryType {
+                    limits: Limits {
+                        min: mem.limits().min(),
+                        max: mem.limits().max(),
+                    },
+                });
+                self.module.section(&memories);
+            }
+            ExternType::Table(table) => {
+                let mut tables = TableSection::new();
+                tables.table(wasm_encoder::TableType {
+                    element_type: wasm_encoder::ValType::FuncRef,
+                    limits: Limits {
+                        min: table.limits().min(),
+                        max: table.limits().max(),
+                    },
+                });
+                self.module.section(&tables);
+            }
+            ExternType::Global(global) => {
+                let mut globals = GlobalSection::new();
+                globals.global(
+                    wasm_encoder::GlobalType {
+                        val_type: wasm_encoder::ValType::I32,
+                        mutable: global.mutability() == Mutability::Var,
+                    },
+                    value_to_instruction(&global.content()),
+                );
+                self.module.section(&globals);
+            }
+            ExternType::Func(ty) => {
+                let mut types = TypeSection::new();
+                types.function(
+                    ty.params().into_iter().map(|it| value_to_value(&it)),
+                    ty.results().into_iter().map(|it| value_to_value(&it)),
+                );
+                let mut functions = FunctionSection::new();
+                functions.function(0);
+
+                let locals = vec![];
+                let mut func = Function::new(locals);
+                for ty in ty.results() {
+                    func.instruction(value_to_instruction(&ty));
+                }
+                let mut codes = CodeSection::new();
+                codes.function(&func);
+                self.module.section(&types);
+                self.module.section(&functions);
+                self.module.section(&codes);
```

Comment:
If `ExternType::Func` is ever provided more than once, then this will define multiple types, functions, and codes sections, but each section must only appear once in a Wasm binary. Similar for other methods. Rather than defining the sections inside this method, and adding them to the module at the end of it, I think the `WencoderGenerator` should have `self.types`, `self.functions`, etc members that have each section, and then `finish` would add them to `self.module` in the [correct order](https://webassembly.github.io/spec/core/binary/modules.html#sections) before calling `self.module.finish()`.

Reply:
@fitzgen  I update the PR. There is one question. I cannot find instruction section in [correct order](https://webassembly.github.io/spec/core/binary/modules.html#sections). So I did not add it to module. Is it right?